File: docs/sharepoint/codesnippet/VisualBasic/sitecolumnprojectitem/sharepointcommands/commands.vb
'<Snippet9>
Imports Microsoft.SharePoint
Imports Microsoft.VisualStudio.SharePoint.Commands

Namespace Contoso.SharePoint.Commands

    Friend Class Commands

        <SharePointCommand(CommandIds.ValidateSite)> _
        Private Function ValidateSite(ByVal context As ISharePointCommandContext, ByVal url As Uri) As Boolean
            Using site As SPSite = New SPSite(url.AbsoluteUri)
                Dim webUrl As String = DetermineWebUrl(url.AbsolutePath, site.ServerRelativeUrl)
                If webUrl IsNot Nothing Then
                    Using web As SPWeb = site.OpenWeb(webUrl, True)
                        Return web.Exists
                    End Using
                End If
            End Using
            Return False
        End Function

        ' For simplicity, this command does not check to make sure the provided Uri is valid. 
        ' Use the ValidateSite command to verify that the Uri is valid first.
        <SharePointCommand(CommandIds.GetFieldTypes)> _
        Private Function GetFieldTypes(ByVal context As ISharePointCommandContext, ByVal url As Uri) As String()
            Dim columnDefinitions As List(Of String) = New List(Of String)()
            Using site As SPSite = New SPSite(url.AbsoluteUri)
                Dim webUrl As String = DetermineWebUrl(url.AbsolutePath, site.ServerRelativeUrl)
                Using web As SPWeb = site.OpenWeb(webUrl, True)
                    For Each columnDefinition As SPFieldTypeDefinition In web.FieldTypeDefinitionCollection
                        columnDefinitions.Add(columnDefinition.TypeName)
                    Next
                    ' SharePoint commands cannot serialize List<string>, so return an array.
                    Return columnDefinitions.ToArray()
                End Using
            End Using
        End Function

        Private Function DetermineWebUrl(ByVal serverRelativeInputUrl As String, ByVal serverRelativeSiteUrl As String) As String
            ' Make sure both URLs have a trailing slash.
            serverRelativeInputUrl = EnsureTrailingSlash(serverRelativeInputUrl)
            serverRelativeSiteUrl = EnsureTrailingSlash(serverRelativeSiteUrl)

            Dim webUrl As String = Nothing
            Dim isSubString As Boolean = serverRelativeInputUrl.StartsWith(serverRelativeSiteUrl, StringComparison.OrdinalIgnoreCase)

            If isSubString Then
                ' The Web URL cannot have escaped characters.
                webUrl = Uri.UnescapeDataString(serverRelativeInputUrl.Substring(serverRelativeSiteUrl.Length))
            End If
            Return webUrl
        End Function

        Private Function EnsureTrailingSlash(ByVal url As String)
            If Not String.IsNullOrEmpty(url) AndAlso url(url.Length - 1) <> "/" Then
                url += "/"
            End If
            Return url
        End Function
    End Class
End Namespace
'</Snippet9>

File: docs/sharepoint/codesnippet/CSharp/sitecolumnprojectitem/sharepointcommands/commands.cs
//<Snippet9>
using System;
using System.Collections.Generic;
using Microsoft.SharePoint;
using Microsoft.VisualStudio.SharePoint.Commands;

namespace Contoso.SharePoint.Commands
{
    internal class Commands
    {
        [SharePointCommand(CommandIds.ValidateSite)]
        private bool ValidateSite(ISharePointCommandContext context, Uri url)
        {
            using (SPSite site = new SPSite(url.AbsoluteUri))
            {
                string webUrl = DetermineWebUrl(url.AbsolutePath, site.ServerRelativeUrl);
                if (webUrl != null)
                {
                    using (SPWeb web = site.OpenWeb(webUrl, true))
                    {
                        return web.Exists;
                    }
                }
            }

            return false;
        }

        // For simplicity, this command does not check to make sure the provided Uri is valid. 
        // Use the ValidateSite command to verify that the Uri is valid first.
        [SharePointCommand(CommandIds.GetFieldTypes)]
        private string[] GetFieldTypes(ISharePointCommandContext context, Uri url)
        {
            List<string> columnDefinitions = new List<string>();
            using (SPSite site = new SPSite(url.AbsoluteUri))
            {
                string webUrl = DetermineWebUrl(url.AbsolutePath, site.ServerRelativeUrl);
                using (SPWeb web = site.OpenWeb(webUrl, true))
                {
                    foreach (SPFieldTypeDefinition columnDefinition in web.FieldTypeDefinitionCollection)
                    {
                        columnDefinitions.Add(columnDefinition.TypeName);
                    }

                    // SharePoint commands cannot serialize List<string>, so return an array.
                    return columnDefinitions.ToArray();
                }
            }
        }

        private string DetermineWebUrl(string serverRelativeInputUrl, string serverRelativeSiteUrl)
        {
            // Make sure both URLs have a trailing slash.
            serverRelativeInputUrl = EnsureTrailingSlash(serverRelativeInputUrl);
            serverRelativeSiteUrl = EnsureTrailingSlash(serverRelativeSiteUrl);

            string webUrl = null;
            bool isSubString = serverRelativeInputUrl.StartsWith(serverRelativeSiteUrl, StringComparison.OrdinalIgnoreCase);

            if (isSubString)
            {
                // The Web URL cannot have escaped characters.
                webUrl = Uri.UnescapeDataString(serverRelativeInputUrl.Substring(serverRelativeSiteUrl.Length));
            }

            return webUrl;
        }

        private string EnsureTrailingSlash(string url)
        {
            if (!String.IsNullOrEmpty(url)
                && url[url.Length - 1] != '/')
            {
                url += '/';
            }
            return url;
        }
    }
}
//</Snippet9>
